File: app/src/main/java/com/suzei/racoon/fragment/AddWorldFragment.java
package com.suzei.racoon.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.suzei.racoon.R;
import com.suzei.racoon.adapter.SearchAdapter;
import com.suzei.racoon.adapter.SelectedAdapter;
import com.suzei.racoon.model.Users;
import com.suzei.racoon.ui.search.SearchContract;
import com.suzei.racoon.ui.search.SearchPresenter;
import com.suzei.racoon.ui.search.SelectPresenter;
import com.suzei.racoon.ui.search.SelectUserContract;
import com.suzei.racoon.util.EmptyRecyclerView;

import java.util.ArrayList;

import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import timber.log.Timber;

public class AddWorldFragment extends Fragment implements
        SearchContract.SearchView,
        SelectUserContract.SelectUserView {

    private SearchPresenter searchPresenter;
    private SelectPresenter selectPresenter;

    private ArrayList<Users> selectedUserList = new ArrayList<>();

    @BindString(R.string.default_user_image) String defaultImage;
    @BindView(R.id.pick_user_no_result) TextView noResultView;
    @BindView(R.id.pick_user_loading) ProgressBar progressBarView;
    @BindView(R.id.pick_user_back) ImageButton backView;
    @BindView(R.id.pick_user_search_text) EditText searchUserView;
    @BindView(R.id.pick_user_search) ImageButton searchView;
    @BindView(R.id.pick_user_selected_list_layout) RelativeLayout userSelectedLayout;
    @BindView(R.id.pick_user_selected_list) RecyclerView listSelectedUserView;
    @BindView(R.id.pick_user_add) Button addView;
    @BindView(R.id.pick_user_search_list) EmptyRecyclerView listSearchUserView;

    public AddWorldFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.pick_user, container, false);
        initObjects(view);
        setUpRecyclerViews();
        setUpPresenters();
        return view;
    }

    private void initObjects(View view) {
        ButterKnife.bind(this, view);
    }

    private void setUpRecyclerViews() {
        listSearchUserView.setLayoutManager(new LinearLayoutManager(getContext()));
        listSelectedUserView.setLayoutManager(new LinearLayoutManager(getContext(),
                LinearLayoutManager.HORIZONTAL, false));
        listSearchUserView.setEmptyView(noResultView);
    }

    private void setUpPresenters() {
        searchPresenter = new SearchPresenter(this);
        selectPresenter = new SelectPresenter(this);
    }

    @OnClick(R.id.pick_user_back)
    public void onBackClick() {
        getActivity().finish();
    }

    @OnClick(R.id.pick_user_search)
    public void onSearchClick() {
        String query = searchUserView.getText().toString();
        searchPresenter.startSearch(query, selectedUserList);
    }

    @Override
    public void showProgress() {

    }

    @Override
    public void hideProgress() {

    }

    @Override
    public void setSearchAdapter(SearchAdapter searchAdapter) {
        listSearchUserView.setAdapter(searchAdapter);
    }

    @Override
    public void setSearchUserItemClick(Users users) {
        selectedUserList.add(users);
        selectPresenter.addSelectedUser(users);
        searchPresenter.removeFromSearch(users);

        listSelectedUserView.scrollToPosition(selectPresenter.getItemCount() - 1);
        userSelectedLayout.setVisibility(View.VISIBLE);
    }

    @Override
    public void searchFailed() {
        Timber.i("Failed");
    }

    @Override
    public void setSelectUserAdapter(SelectedAdapter selectUserAdapter) {
        listSelectedUserView.setAdapter(selectUserAdapter);
    }

    @Override
    public void setSelectUserItemClick(Users users) {
        selectedUserList.remove(users);
        selectPresenter.removeSelectedUser(users);
        searchPresenter.addFromSearch(users);

        if (selectedUserList.size() == 0) {
            userSelectedLayout.setVisibility(View.GONE);
        }

    }

    @Override
    public void selectUserFailed() {
        Timber.i("Failed");
    }
}

File: app/src/main/java/com/suzei/racoon/ui/add/AddWorldFragment.kt
package com.suzei.racoon.ui.add

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.database.FirebaseDatabase
import com.suzei.racoon.databinding.PickUserBinding
import com.suzei.racoon.model.Users
import com.suzei.racoon.ui.add.create.CreateGroupContract.CreateGroupView
import com.suzei.racoon.ui.add.create.CreateGroupPresenter
import com.suzei.racoon.ui.add.search.*
import com.suzei.racoon.ui.add.search.SelectUserContract.SelectUserView
import com.suzei.racoon.ui.chatroom.group.GroupChatActivity
import timber.log.Timber
import java.util.*

class AddWorldFragment : Fragment(), SearchContract.SearchView, SelectUserView, CreateGroupView {
    private var searchPresenter: SearchPresenter? = null
    private var selectPresenter: SelectPresenter? = null
    private var createGroupPresenter: CreateGroupPresenter? = null
    private val selectedUserList = ArrayList<Users>()

    private var _binding: PickUserBinding? = null
    private val binding get() = _binding!!

    private val onBackClick = View.OnClickListener {
        requireActivity().finish()
    }

    private val onAddClick = View.OnClickListener {
        createGroupPresenter!!.create(selectedUserList)
    }

    private val onSearchClick = View.OnClickListener {
        val query = binding.pickUserSearchText.text.toString()
        searchPresenter!!.startSearch(query, selectedUserList)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = PickUserBinding.inflate(layoutInflater)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setUpRecyclerViews()
        setUpPresenters()
        setupClickListener()
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }

    private fun setUpRecyclerViews() {
        binding.pickUserSearchList.layoutManager = LinearLayoutManager(context)
        binding.pickUserSelectedList.layoutManager = LinearLayoutManager(
            context,
            LinearLayoutManager.HORIZONTAL, false
        )
        binding.pickUserSearchList.setEmptyView(binding.pickUserNoResult)
    }

    private fun setUpPresenters() {
        searchPresenter = SearchPresenter(this)
        selectPresenter = SelectPresenter(this)
        createGroupPresenter =
            CreateGroupPresenter(context, this)
    }

    private fun setupClickListener() {
        binding.pickUserBack.setOnClickListener(onBackClick)
        binding.pickUserAdd.setOnClickListener(onAddClick)
        binding.pickUserSearch.setOnClickListener(onSearchClick)
    }

    override fun createSuccess(id: String) {
        FirebaseDatabase.getInstance().reference.child("world_chats").child(id).setValue(true)
        val chatRoomIntent = Intent(context, GroupChatActivity::class.java)
        chatRoomIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        chatRoomIntent.putExtra(GroupChatActivity.EXTRA_GROUP_ID, id)
        startActivity(chatRoomIntent)
        requireActivity().finish()
    }

    override fun createFailed() {}
    override fun setSearchAdapter(searchAdapter: SearchAdapter) {
        binding.pickUserSearchList.adapter = searchAdapter
    }

    override fun setSearchUserItemClick(users: Users) {
        selectedUserList.add(users)
        selectPresenter!!.addSelectedUser(users)
        searchPresenter!!.removeFromSearch(users)
        binding.pickUserSelectedList.scrollToPosition(selectPresenter!!.itemCount - 1)
        binding.pickUserSelectedListLayout.visibility = View.VISIBLE
    }

    override fun searchSuccess() {}
    override fun searchFailed() {
        Timber.i("Failed")
    }

    override fun showProgress() {
        binding.pickUserSearch.isEnabled = false
        binding.pickUserLoading.visibility = View.VISIBLE
    }

    override fun hideProgress() {
        binding.pickUserSearch.isEnabled = true
        binding.pickUserLoading.visibility = View.GONE
    }

    override fun setSelectUserAdapter(selectUserAdapter: SelectedAdapter) {
        binding.pickUserSelectedList.adapter = selectUserAdapter
    }

    override fun setSelectUserItemClick(users: Users) {
        selectedUserList.remove(users)
        selectPresenter!!.removeSelectedUser(users)
        searchPresenter!!.addFromSearch(users)
        if (selectedUserList.size == 0) {
            binding.pickUserSelectedListLayout.visibility = View.GONE
        }
    }

    override fun selectUserFailed() {
        Timber.i("Failed")
    }
}